test/identifier/single-statement.spec.js

```diff
@@ -97,12 +97,39 @@ describe('identifier', function () {
     });
 
     it('should identify postgres "CREATE FUNCTION" statement', function () {
-      const actual = identify("CREATE FUNCTION dup(in int, out f1 int, out f2 text) AS $$ SELECT $1, CAST($1 AS text) || ' is text' $$ LANGUAGE SQL;");
+      const sql = `CREATE FUNCTION quarterly_summary_func(start_date date DEFAULT CURRENT_TIMESTAMP)
+      RETURNS TABLE (staff_name text, staff_bonus int, quarter tsrange)
+      As $$
+      DECLARE
+        employee RECORD;
+        total_bonus int;
+        sales_total int;
+        end_date date := start_date + interval '3 months';
+      BEGIN
+        FOR employee IN SELECT staff_id FROM staff LOOP
+          EXECUTE 'SELECT sum(staff_bonus), sum(sales_price) FROM sales WHERE staff_id = $1
+          AND created_at >= $2 AND created_at < $3'
+             INTO total_bonus, sales_total
+             USING employee.staff_id, start_date, end_date;
+          RAISE NOTICE 'total bonus is % and total sales is %', total_bonus, sales_total;
+         EXECUTE 'INSERT INTO sales_summary (staff_id, bonus, total_sales, period) VALUES
+                    ($1, $2, $3, tsrange($4, $5))'
+             USING employee.staff_id, total_bonus, sales_total, start_date, end_date;
+        END LOOP;
+        DELETE FROM sales WHERE created_at >= start_date
+               AND created_at < end_date;
+        RETURN QUERY SELECT name, bonus, period FROM sales_summary
+                     LEFT JOIN staff on sales_summary.staff_id = staff.staff_id;
+       RETURN;
+      END;
+      $$
+      LANGUAGE plpgsql;`;
```

Comment:
I think unfortunately it's valid for `LANGUAGE` to be before the `BEGIN` block, which would throw out the logic of your parser.

For example this query from https://stackoverflow.com/questions/8918755/postgresql-function-with-a-loop
switching `LANGUAGE` to before `AS` it is still valid:

```sql
CREATE OR REPLACE FUNCTION f_grp_prod(text)
  RETURNS TABLE (
    name text
  , result1 double precision
  , result2 double precision)
LANGUAGE plpgsql STABLE
AS
$BODY$
DECLARE
    r      mytable%ROWTYPE;
    _round integer;
BEGIN
    -- init vars
    name    := $1;
    result2 := 1;       -- abuse result2 as temp var for convenience

FOR r IN
    SELECT *
    FROM   mytable m
    WHERE  m.name = name
    ORDER  BY m.round
LOOP
    IF r.round <> _round THEN   -- save result1 before 2nd round
        result1 := result2;
        result2 := 1;
    END IF;

    result2 := result2 * (1 - r.val/100);
    _round  := r.round;
END LOOP;

RETURN NEXT;

END;
$BODY$;

```

See for example - https://www.postgresqltutorial.com/postgresql-create-function/

Reply:
You are absolutely right. I've included this function as another test case, and modified the code to account for this, adjusting it to fit my latter thought on how to accomplish this (keeping track of block openers).